### PR TITLE
refactor(add-extension): Update add-extension.js

### PR DESCRIPTION
**Pull request by vymarkov**
Export a few functions in order to make possible usage of this module by external packages, it will allow to create external tools to automate process of migration extension packs that usually contain a bunch of not published extensions to Open VSX.

Use case: 
```js
const { addNewExtension, onDidAddExtension, writeToExtensionsFile, readExtensionsFromFile, fetchExtInfoFromClonedRepo } = require('./add-extension')

const start = async () => {
    const extensionsFile = './extensions.json'
    const { extensions } = await readExtensionsFromFile(extensionsFile)
    const repos = ['https://github.com/oguhpereira/vscode-spotify-theme', 'https://github.com/whizkydee/vscode-palenight-theme']

    const addedExtensions = []
    for (const repository of repos) {
        const { extension, package } = await fetchExtInfoFromClonedRepo(repository, {})
        await addNewExtension(extension, package, extensions);
        addedExtensions.push(extension)
    }
    await writeToExtensionsFile(extensions, extensionsFile)
    addedExtensions.forEach(extension => onDidAddExtension(extension))
}

start()
```

### CODE REVIEW

**Comment by jankeromnes** [comment]
Many thanks for proposing to extract useful functions into re-usable modules! I very much like this idea, please let me know if I can help.

Also, I notice that this Pull Request includes merge commits and unrelated commits. When you get a chance, could you please rebase your branch on top of the latest `master` state? (Note: GitHub's "Update Pull Request" button will always use merge commits, so it can't be used)

**Comment by vymarkov** [comment]
I rebased the branch, looks like it's ok now :) 

I'm working on cli util that allows to validate extension pack by name and optionally add some of them (that have licence that is OSI approved) to `./extensions.json`

```bash
# (dont worry, it takes around 90 seconds, it's ok, this cli is under active development)
$ npm i -g vymarkov/validate-vscode-extension-pack 
``` 

currently I have one blocker that doesn't allow to use this cli without any restrictions to add extension pack itself to `./extensions.json`, otherwise extension pack can't be added to `./extensions.json` file until all extensions are present Open VSX registry.

https://github.com/eclipse/openvsx/issues/224#issuecomment-755140689 

@jankeromnes it would be nice if you can help with this issue

**Comment by vymarkov** [comment]
There is an example with `mubaidr.vuejs-extension-pack`, you can find that two extensions can not be published to Open VSX registry due to licence issue, so it would be nice to have ability to publish packs even if some extensions are not published.

```bash
$ validate-vscode-extension-pack mubaidr.vuejs-extension-pack ./extensions.json 
```

Output:
```bash
See below extensions that are not present in Open VSX marketplace:
extensions with licence (1): [
  {
    name: 'sdras.vue-vscode-snippets',
    msmarketplace: 'https://marketplace.visualstudio.com/items?itemName=sdras.vue-vscode-snippets',
    openvsx: 'https://open-vsx.org/extension/sdras/vue-vscode-snippets',
    repoUrl: 'https://github.com/sdras/vue-vscode-snippets.git',
    licence: 'MIT',
    licenceUrl: null,
    lastUpdated: '2020-09-02T19:57:07.793Z'
  }
]
extensions without licence (2): [
  {
    name: 'formulahendry.auto-complete-tag',
    msmarketplace: 'https://marketplace.visualstudio.com/items?itemName=formulahendry.auto-complete-tag',
    openvsx: 'https://open-vsx.org/extension/formulahendry/auto-complete-tag',
    repoUrl: 'https://github.com/formulahendry/vscode-auto-complete-tag.git',
    licence: null,
    licenceUrl: null,
    lastUpdated: '2018-08-17T08:53:57.377Z'
  },
  {
    name: 'visualstudioexptteam.vscodeintellicode',
    msmarketplace: 'https://marketplace.visualstudio.com/items?itemName=visualstudioexptteam.vscodeintellicode',
    openvsx: 'https://open-vsx.org/extension/visualstudioexptteam/vscodeintellicode',
    repoUrl: 'https://github.com/MicrosoftDocs/intellicode',
    licence: 'NOASSERTION',
    licenceUrl: null,
    lastUpdated: '2020-08-11T23:58:56.807Z'
  }
]
Extensions without licence CAN NOT BE uploaded to Open VSX registry, LICENCE must be present
```

**Comment by vymarkov** [comment]
@jankeromnes there is another example of usage cli util that reuse the exported functions from `add-extension.js`, in the near future I'd would like to see the migration of extension pack that contains lots of unpublished extensions in this way: 

```bash
$ git clone https://github.com/open-vsx/publish-extensions # or fork and clone repo
$ npm i -g https://github.com/vymarkov/validate-vscode-extension-pack # I'm gonna rename it :) 
$ ext-utils add --add-extensions-with-license --itself jabacchetta.vscode-essentials ./extensions.json
```

Output
```bash
Some of extensions are not open source thus can not published to Open VSX from https://github.com/open-vsx/publish-extensions
You need to ask authors to publish extension by himself:
https://marketplace.visualstudio.com/items?itemName=visualstudioexptteam.vscodeintellicode
 
 
See below extensions that are not present in Open VSX marketplace:
extensions with license (4): [
  {
    name: 'jkjustjoshing.vscode-text-pastry',
    msmarketplaceUrl: 'https://marketplace.visualstudio.com/items?itemName=jkjustjoshing.vscode-text-pastry',
    openvsx: 'https://open-vsx.org/extension/jkjustjoshing/vscode-text-pastry',
    repoUrl: 'https://github.com/jkjustjoshing/vscode-text-pastry',
    license: 'MIT',
    licenseUrl: null,
    lastUpdated: '2017-10-29T00:40:41.967Z'
  },
  {
    name: 'lacroixdavid1.vscode-format-context-menu',
    msmarketplaceUrl: 'https://marketplace.visualstudio.com/items?itemName=lacroixdavid1.vscode-format-context-menu',
    openvsx: 'https://open-vsx.org/extension/lacroixdavid1/vscode-format-context-menu',
    repoUrl: 'https://github.com/lacroixdavid1/vscode-format-context-menu',
    license: 'MIT',
    licenseUrl: null,
    lastUpdated: '2018-09-06T20:57:12.447Z'
  },
  {
    name: 'jasonlhy.hungry-delete',
    msmarketplaceUrl: 'https://marketplace.visualstudio.com/items?itemName=jasonlhy.hungry-delete',
    openvsx: 'https://open-vsx.org/extension/jasonlhy/hungry-delete',
    repoUrl: 'https://github.com/Jasonlhy/VSCode-Hungry-Delete',
    license: 'MIT',
    licenseUrl: null,
    lastUpdated: '2019-04-22T17:58:17.747Z'
  },
  {
    name: 'inu1255.easy-snippet',
    msmarketplaceUrl: 'https://marketplace.visualstudio.com/items?itemName=inu1255.easy-snippet',
    openvsx: 'https://open-vsx.org/extension/inu1255/easy-snippet',
    repoUrl: 'https://github.com/inu1255/vscode-easy-snippet',
    license: 'MIT',
    licenseUrl: null,
    lastUpdated: '2020-08-31T02:44:37.873Z'
  }
]
extensions with not identified license or license is not present (14): [
  {
    name: 'hoovercj.vscode-settings-cycler',
    msmarketplaceUrl: 'https://marketplace.visualstudio.com/items?itemName=hoovercj.vscode-settings-cycler',
    openvsx: 'https://open-vsx.org/extension/hoovercj/vscode-settings-cycler',
    repoUrl: 'https://www.github.com/hoovercj/vscode-api-playground',
    license: null,
    licenseUrl: null,
    lastUpdated: '2017-04-13T21:29:39.703Z'
  },
  {
    name: 'slevesque.vscode-multiclip',
    msmarketplaceUrl: 'https://marketplace.visualstudio.com/items?itemName=slevesque.vscode-multiclip',
    openvsx: 'https://open-vsx.org/extension/slevesque/vscode-multiclip',
    repoUrl: 'https://github.com/stef-levesque/vscode-multiclip',
    license: null,
    licenseUrl: null,
    lastUpdated: '2018-10-09T04:49:32.11Z'
  },
  {
    name: 'redvanworkshop.explorer-exclude-vscode-extension',
    msmarketplaceUrl: 'https://marketplace.visualstudio.com/items?itemName=redvanworkshop.explorer-exclude-vscode-extension',
    openvsx: 'https://open-vsx.org/extension/redvanworkshop/explorer-exclude-vscode-extension',
    repoUrl: 'https://github.com/redvanworkshop/explorer-exclude-vscode-extension',
    license: null,
    licenseUrl: null,
    lastUpdated: '2019-07-06T10:04:39.94Z'
  },
  {
    name: 'pnp.polacode',
    msmarketplaceUrl: 'https://marketplace.visualstudio.com/items?itemName=pnp.polacode',
    openvsx: 'https://open-vsx.org/extension/pnp/polacode',
    repoUrl: 'https://github.com/octref/polacode',
    license: null,
    licenseUrl: null,
    lastUpdated: '2019-09-27T16:24:31.86Z'
  },
  {
    name: 'jakearl.search-editor-apply-changes',
    msmarketplaceUrl: 'https://marketplace.visualstudio.com/items?itemName=jakearl.search-editor-apply-changes',
    openvsx: 'https://open-vsx.org/extension/jakearl/search-editor-apply-changes',
    repoUrl: 'https://github.com/JacksonKearl/vscode-search-editor-apply-changes',
    license: null,
    licenseUrl: null,
    lastUpdated: '2020-03-03T23:45:34.57Z'
  },
  {
    name: 'micnil.vscode-checkpoints',
    msmarketplaceUrl: 'https://marketplace.visualstudio.com/items?itemName=micnil.vscode-checkpoints',
    openvsx: 'https://open-vsx.org/extension/micnil/vscode-checkpoints',
    repoUrl: 'https://github.com/micnil/vscode-checkpoints',
    license: null,
    licenseUrl: null,
    lastUpdated: '2020-05-17T11:10:38.817Z'
  },
  {
    name: 'metaseed.metago',
    msmarketplaceUrl: 'https://marketplace.visualstudio.com/items?itemName=metaseed.metago',
    openvsx: 'https://open-vsx.org/extension/metaseed/metago',
    repoUrl: 'https://github.com/metaseed/metaGo',
    license: null,
    licenseUrl: null,
    lastUpdated: '2020-06-09T21:04:38.113Z'
  },
  {
    name: 'tomoki1207.pdf',
    msmarketplaceUrl: 'https://marketplace.visualstudio.com/items?itemName=tomoki1207.pdf',
    openvsx: 'https://open-vsx.org/extension/tomoki1207/pdf',
    repoUrl: 'https://github.com/tomoki1207/vscode-pdfviewer',
    license: null,
    licenseUrl: null,
    lastUpdated: '2020-07-13T00:32:36.377Z'
  },
  {
    name: 'karigari.chat',
    msmarketplaceUrl: 'https://marketplace.visualstudio.com/items?itemName=karigari.chat',
    openvsx: 'https://open-vsx.org/extension/karigari/chat',
    repoUrl: 'https://github.com/vsls-contrib/chat',
    license: null,
    licenseUrl: null,
    lastUpdated: '2020-09-07T16:55:55.717Z'
  },
  {
    name: 'ms-vsliveshare.vsliveshare-audio',
    msmarketplaceUrl: 'https://marketplace.visualstudio.com/items?itemName=ms-vsliveshare.vsliveshare-audio',
    openvsx: 'https://open-vsx.org/extension/ms-vsliveshare/vsliveshare-audio',
    repoUrl: 'https://github.com/MicrosoftDocs/live-share',
    license: 'NOASSERTION',
    licenseUrl: null,
    lastUpdated: '2020-09-15T19:14:21.957Z'
  },
  {
    name: 'nhoizey.gremlins',
    msmarketplaceUrl: 'https://marketplace.visualstudio.com/items?itemName=nhoizey.gremlins',
    openvsx: 'https://open-vsx.org/extension/nhoizey/gremlins',
    repoUrl: 'https://github.com/nhoizey/vscode-gremlins.git',
    license: null,
    licenseUrl: null,
    lastUpdated: '2020-11-05T22:10:52.25Z'
  },
  {
    name: 'lostintangent.vsls-whiteboard',
    msmarketplaceUrl: 'https://marketplace.visualstudio.com/items?itemName=lostintangent.vsls-whiteboard',
    openvsx: 'https://open-vsx.org/extension/lostintangent/vsls-whiteboard',
    repoUrl: 'https://github.com/lostintangent/vsls-whiteboard',
    license: null,
    licenseUrl: null,
    lastUpdated: '2020-12-09T03:23:54.507Z'
  },
  {
    name: 'vsls-contrib.gistfs',
    msmarketplaceUrl: 'https://marketplace.visualstudio.com/items?itemName=vsls-contrib.gistfs',
    openvsx: 'https://open-vsx.org/extension/vsls-contrib/gistfs',
    repoUrl: 'https://github.com/vsls-contrib/gistpad',
    license: null,
    licenseUrl: null,
    lastUpdated: '2020-12-29T00:22:50.797Z'
  },
  {
    name: 'codestream.codestream',
    msmarketplaceUrl: 'https://marketplace.visualstudio.com/items?itemName=codestream.codestream',
    openvsx: 'https://open-vsx.org/extension/codestream/codestream',
    repoUrl: 'https://github.com/TeamCodeStream/CodeStream',
    license: null,
    licenseUrl: null,
    lastUpdated: '2021-01-19T23:34:23.45Z'
  }
]
Extensions without license CAN NOT BE uploaded to Open VSX registry, license must be present

Adding extensions with defined license to ./extensions.json
Running: git clone --recurse-submodules https://github.com/jkjustjoshing/vscode-text-pastry /tmp/repository
Running: ls package.json 2>/dev/null || git ls-files | grep package\.json
Running: rm -rf /tmp/repository
[WARNING] Could not check Open VSX version of jkjustjoshing.vscode-text-pastry:
Error: Extension not found: jkjustjoshing.vscode-text-pastry
    at IncomingMessage.<anonymous> (/home/gitpod/.nvm/versions/node/v14.15.4/lib/node_modules/extpack/node_modules/ovsx/lib/registry.js:173:40)
    at IncomingMessage.emit (events.js:327:22)
    at endReadableNT (internal/streams/readable.js:1327:12)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)
Running: git clone --recurse-submodules https://github.com/lacroixdavid1/vscode-format-context-menu /tmp/repository
Running: ls package.json 2>/dev/null || git ls-files | grep package\.json
Running: rm -rf /tmp/repository
[WARNING] Could not check Open VSX version of lacroixdavid1.vscode-format-context-menu:
Error: Extension not found: lacroixdavid1.vscode-format-context-menu
    at IncomingMessage.<anonymous> (/home/gitpod/.nvm/versions/node/v14.15.4/lib/node_modules/extpack/node_modules/ovsx/lib/registry.js:173:40)
    at IncomingMessage.emit (events.js:327:22)
    at endReadableNT (internal/streams/readable.js:1327:12)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)
Running: git clone --recurse-submodules https://github.com/Jasonlhy/VSCode-Hungry-Delete /tmp/repository
Running: ls package.json 2>/dev/null || git ls-files | grep package\.json
Running: rm -rf /tmp/repository
[WARNING] Could not check Open VSX version of jasonlhy.hungry-delete:
Error: Extension not found: jasonlhy.hungry-delete
    at IncomingMessage.<anonymous> (/home/gitpod/.nvm/versions/node/v14.15.4/lib/node_modules/extpack/node_modules/ovsx/lib/registry.js:173:40)
    at IncomingMessage.emit (events.js:327:22)
    at endReadableNT (internal/streams/readable.js:1327:12)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)
Running: git clone --recurse-submodules https://github.com/inu1255/vscode-easy-snippet /tmp/repository
Running: ls package.json 2>/dev/null || git ls-files | grep package\.json
Running: rm -rf /tmp/repository
[WARNING] Could not check Open VSX version of inu1255.easy-snippet:
Error: Extension not found: inu1255.easy-snippet
    at IncomingMessage.<anonymous> (/home/gitpod/.nvm/versions/node/v14.15.4/lib/node_modules/extpack/node_modules/ovsx/lib/registry.js:173:40)
    at IncomingMessage.emit (events.js:327:22)
    at endReadableNT (internal/streams/readable.js:1327:12)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)
[OK] Succesfully added new extension: {
  "id": "jkjustjoshing.vscode-text-pastry",
  "repository": "https://github.com/jkjustjoshing/vscode-text-pastry",
  "version": "1.2.0"
}
[OK] Succesfully added new extension: {
  "id": "lacroixdavid1.vscode-format-context-menu",
  "repository": "https://github.com/lacroixdavid1/vscode-format-context-menu",
  "version": "1.0.4"
}
[OK] Succesfully added new extension: {
  "id": "jasonlhy.hungry-delete",
  "repository": "https://github.com/Jasonlhy/VSCode-Hungry-Delete",
  "version": "1.6.0"
}
[OK] Succesfully added new extension: {
  "id": "inu1255.easy-snippet",
  "repository": "https://github.com/inu1255/vscode-easy-snippet",
  "version": "0.5.0"
}
```

It works pretty well, but this util has some drawbacks, the main is `add-extension.js` doesn't provide ability to check the available tags for an extension and use one of them as value for `checkout` field, as it described inside [readme](https://github.com/open-vsx/publish-extensions/blob/master/README.md), I'd like to implement it on the separate pull request, this PR is large enough :)

**Comment by vymarkov** [comment]
@spoenemann could you take a look at this PR? 

Thank you in advance.

File: add-extension.js [hunk 1]
```diff
@@ -20,131 +20,150 @@ const { DH_UNABLE_TO_CHECK_GENERATOR } = require('constants');
...
+      process.exit();
+    }
+  
+    // Handle 'node add-extension --download=VSIX_URL':
+    if (argv.download) {
+      try {
+        await exec('mkdir -p /tmp/vsix');
+        await exec(`wget -O extension.vsix ${argv.download}`, { cwd: '/tmp/vsix' });
+        await exec('unzip -q extension.vsix', { cwd: '/tmp/vsix' });
+        /** @type {{ publisher: string, name: string, version: string }} */
```
Comment (spoenemann) [inline add-extension.js:66]: The `ovsx` library should include type information, so this type comment can be removed.

File: add-extension.js [hunk 2]
```diff
@@ -20,131 +20,150 @@ const { DH_UNABLE_TO_CHECK_GENERATOR } = require('constants');
...
-Alternative usage: node add-extension --download=VSIX_URL`);
-    process.exitCode = 1;
-    process.exit();
-  }
-
-  // Handle 'node add-extension --download=VSIX_URL':
-  if (argv.download) {
+// don't run this part if this module is not exported 
+// @ts-ignore
+if (!module.parent) {
```
Comment (spoenemann) [inline add-extension.js:27]: What is the exact meaning of this? Could you explain a bit more in the comment?

**Review by spoenemann — COMMENTED** [review]
Yes, but it would be good to have @jankeromnes test this as well. I have not been involved in this part of the code yet.

File: add-extension.js (hunk 2 above)
Comment (vymarkov) [reply add-extension.js:27]: @spoenemann it allows us to run the target code of `add-extension.js` if we don't want to use some function from this module, there is [an example](https://github.com/vymarkov/validate-vscode-extension-pack/blob/main/src/cli.ts#L7), I think, it better to extract the functions into file with name `utils.js` and then import them.

Comment (spoenemann) [reply add-extension.js:27]: Yes, better separate library code (reusable modules) from executable code. Could you move the functions you'd like to reuse elsewhere into a new file in the `lib` subdirectory?

Comment (vymarkov) [reply add-extension.js]: @spoenemann @jankeromnes I've just extracted the functions into lib directory

File: add-extension.js (hunk 1 above)
Comment (vymarkov) [reply add-extension.js]: @spoenemann fixed

**Comment by jankeromnes** [comment]
Hi @vymarkov! I'm so sorry for being unresponsive here, and then merging conflicting changes in https://github.com/open-vsx/publish-extensions/pull/279 ... 🙈

Could you please kindly rebase your changes? 🙏 Then I'd be happy to make another review (with a quicker response time!)

**Comment by vymarkov** [comment]
> Hi @vymarkov! I'm so sorry for being unresponsive here, and then merging conflicting changes in #279 ... 🙈
> 
> Could you please kindly rebase your changes? 🙏 Then I'd be happy to make another review (with a quicker response time!)

Hi @jankeromnes! I've just rebased changes from master, it's ready for review! 
Thanks in advance.

**Comment by vymarkov** [comment]
@jankeromnes hi, I just want to remind you about pending review :)

**Review by jankeromnes — COMMENTED** [review]
Many thanks for the great work, and for the review reminder! 🙏

I added a few thoughts in-line.

In general, I believe I've gotten a bit scared of abstractions, because they make the code a lot more complicated (e.g. more indirections to jump through while trying to understand what the code does), while not always providing clear benefits. So, I now always hard-code and in-line everything to keep the code as simple as possible, and only ever made functions re-usable when there is a clear benefit to it (e.g. at least multiple places need to call the same function).

- Could you please try to only abstract and export functions that have a clear mandate for re-usability? (E.g. I don't think `cloneRepo` should be exported)
- Also, please only make values configurable (e.g. as function parameters) when there is a clear benefit to customizing them. Otherwise, please hard-code the values (e.g. I'm not sure if the ability to provide a different extensions file path outweighs the complexity of requiring everyone to specify `'./extensions.json'` every time they call the read or write function)

And finally, I don't quite like the idea to create a module called `utils` that contains a ton of seemingly unrelated functions. The drawback is that a generic `utils` module tends to accumulate a bunch of unrelated functions, that may start depending on each other's code even when it doesn't quite make sense, and that's a big factor increasing the complexity of the code base.

Instead, could you please try to abstract away specific modules with a clear purpose? E.g. there is already a `exec.js` module, whose purpose is only to execute commands in a specific way. There is now also a `github-scraper.js` module whose purpose is only to get precise informations from GitHub.

Maybe you could refator the generic `utils.js` module into a more specific module like `extensions-manager.js` (with only high-level functions like `addNewExtension(repository)`, letting the module take care of loading `./extensions.json` and saving it when appropriate?)

Just some general thoughts here. Please feel free to disagree! 😅

**Comment by vymarkov** [comment]
> In general, I believe I've gotten a bit scared of abstractions, because they make the code a lot more complicated (e.g. more indirections to jump through while trying to understand what the code does), while not always providing clear benefits. So, I now always hard-code and in-line everything to keep the code as simple as possible, and only ever made functions re-usable when there is a clear benefit to it (e.g. at least multiple places need to call the same high-level function, and it abstracts away a lot of complexity instead of exposing it to the caller via API or function parameters).

Thanks a lot for feedback 👍 Those methods were extracted due to this use case https://github.com/open-vsx/publish-extensions/pull/259#issuecomment-763765815, otherwise it doesn't allow to reuse this methods. 

That was a main reason to extract them into a few functions.

**Comment by akosyakov** [comment]
I am going to close it for now. We refactored quite a lot to allow auto upgrading each night. As a consequence many scripts were removed.

**Pull request closed without merging**